### PR TITLE
kgctl connect

### PR DESCRIPTION
**Pull request by leonnicolas**
Use kgctl connect to connect your laptop to a cluster.

Signed-off-by: leonnicolas <leonloechner@gmx.de>

### CODE REVIEW

**Comment by squat** [comment]
I reviewed the whole PR and it looks pretty good overall. In retrospect, I agree that re-writing the route generation code here is simpler / leads to less hacky code than trying to force the `Route` method to work for Peers. I've been cleaning it up locally and have several fixes / improvements. I'll push them as an additional commit to this branch. However, the code is affected by all of the bug fixes I just made PRs for, so we should probably merge those first and then we can rebase this work on main.

**Comment by squat** [comment]
I have everything working now! Even reusing an existing private-key on disk for a Peer that already exists. I'm trying to get an e2e test working but there is a funny detail: when running this locally, you need sudo to make kgctl have CAP_NET_ADMIN but when running in CI we don't have sudo.

We got around this in the existing ewe tests by using the docker daemon for privileged escalation. This is also possible here, but we don't have any containers with kgctl installed. We can again get around this by running a container and doing "docker cp ..." to  get the binary in there; there are just some tricky details with terminating the container and handling signals correctly.

**Comment by leonnicolas** [comment]
We could run the kgctl connect test  only in CI by enabling it with some env var. and if you really want to test it locally you have to set the var and use sudo?

Or could we use `docker run --network host -v $(pwd)/src golang run src/cmd/kgctl`? But that's a bit messed up to run this on peoples computers that don't know about this as it adds ip routes. Or can we use a new docker network and fo everything in there? I don't much about them

**Comment by squat** [comment]
`docker run --network host -v $(pwd)/src golang run src/cmd/kgctl` ah wow that's genius. This hadn't occurred to me.

We already have ewe tests that use docker to change host networking routes, interfaces, iptables etc: https://github.com/squat/kilo/blob/main/e2e/lib.sh#L181-L200

I think this is _OK_ if you want to contribute and run the tests locally. But I like the idea to add a PR to make the host networking tests optional with a flag.

File: pkg/mesh/topology.go [hunk 1]
```diff
@@ -90,6 +91,21 @@ type segment struct {
 	allowedLocationIPs []net.IPNet
 }
 
+// CIDRS returns the cidrs of the segment.
+func (s Segment) CIDRS() []*net.IPNet {
```
Comment (squat) [inline pkg/mesh/topology.go:95]: Im still a bit on the fence about making these a methods or simply exposing all of the fields of the segment. I think that exposing the fields would be easier + it doesn't make it a "weaker" contract at all. Going forward, all users of this library can expect to be able to find this data from a segment, but using the field directly seems a bit easier to me.

**Comment by squat** [comment]
ok, `go run` didn't work because of the mentioned signal forwarding issue `https://github.com/golang/go/issues/40467` instead I mounted the `$KGCTL_BINARY" path into an alpine container and use it directly. This is straightforward and doesn't involve any of the complicated `docker cp` stuff. Thank goodness for statically compile binaries :heart_eyes: 

e2e tests are working!

**Comment by squat** [comment]
I wonder if we should mention `VPN` somewhere in the command docs / description?? Eventually we should also document this on the main README.md or the vpn.md or in a new doc for connecting to the cluster as a developer trying to work on your services

File: pkg/mesh/topology.go (hunk 1 above)
Comment (leonnicolas) [reply pkg/mesh/topology.go:95]: fine with me to expose them. Or we move the routes function defined in kgctl's main package into the mesh package?

Comment (squat) [reply pkg/mesh/topology.go:95]: Ah interesting. Like a PeerRoutes method? I like that too

Comment (squat) [reply pkg/mesh/topology.go:95]: This way we can have tests

File: docs/kgctl.md [hunk 2]
```diff
@@ -46,9 +46,30 @@ paru -S kgctl-bin
...
+The `connect` command configures the local host as a WireGuard Peer of the cluster and applies all of the necessary networking configuration to connect to the cluster.
+As long as the process is running, it will watch the cluster for changes and automatically manage the configuration for new or updated Peers and Nodes.
+If the given Peer name does not exist in the cluster, the command will register a new Peer and generate the necessary WireGuard keys.
+When the command exits, all of the configuration, including newly registered Peers, is cleaned up.
+
+Example:
+
+```shell
+PEER_NAME=laptop
+kgctl connect $PEER_NAME
```
Comment (leonnicolas) [inline docs/kgctl.md:64]: should we mention that it would be helpful to specify the service cidr as an allowed IP, so you can use IPs of services?

**Comment by leonnicolas** [comment]
> I wonder if we should mention `VPN` somewhere in the command docs / description?? Eventually we should also document this on the main README.md or the vpn.md or in a new doc for connecting to the cluster as a developer trying to work on your services

it would be so helpful to use the cluster's DNS service to test services on your laptop just like if they were running in the cluster

File: docs/kgctl.md (hunk 2 above)
Comment (squat) [reply docs/kgctl.md:64]: Yes 👍

**Comment by squat** [comment]
> it would be so helpful to use the cluster's DNS service to test services on your laptop just like if they were running in the cluster

Yeah i think this would be super helpful in all of our VPN docs. I wonder if we want to automate this with kgctl connect. That would be so helpful.  Like, unless you do --dns=false, we try to modify your resolv.conf? It's sketchy because this is often managed by another service like systemd-resolved

**Comment by leonnicolas** [comment]
>  It's sketchy because this is often managed by another service like systemd-resolved

I think it is out of scope, but we could mention it somewhere

**Comment by squat** [comment]
Agreed. Maybe a plan for the future

File: docs/kgctl.md (hunk 2 above)
Comment (squat) [reply docs/kgctl.md]: done!

**Comment by squat** [comment]
WDYT @leonnicolas? I added the DNS docs and refactored the topology struct to re-simplify it. The PR got like 20% smaller :)

**Review by leonnicolas — COMMENTED** [review]
Very cool. Can't approve because I opened it

**Review by squat — APPROVED** [review]

**Comment by squat** [comment]
I'll prepare the release now :)

**Pull request merged**